docs/config.md
# Config

Not sure what a config value means? You're in the right place, here we'll go over every config option available for zdiscord and what it means and expects. **If you haven't already, please read the setup guide in the [read me](https://github.com/zfbx/zdiscord/blob/djs/readme.md#setup) first.**

**NOTE:** All config values use strings (wrapped in `" "`) for their values. This is done for convar support.

#### Convars
If you don't want to set config options within `config.js` you can set them through the server.cfg as [convars](https://docs.fivem.net/docs/scripting-reference/convars/#standard-convars). If a convar is listed under a config you can use it instead of setting the config as long as it's placed above/before zdiscord is started/ensured or it wont be able to use them. **To use these convars, add a line to your server.cfg near the top with `set convar_name "value"`. replace `convar_name` with the convar name you want to use.

---



# General Configuration Settings

### LanguageLocaleCode
**[ Default: "en" | Convar: `discord_lang` ]**<br>
This determines which locale (translation) file is loaded from `/locales` so if the file that fits the language you want used is in that folder then use the 2 character name of the file. As of writing this ar, de, en, pl, tr and vn are available options.


## Public Values

### FiveMServerName
**[ Replicated | Convar: `discord_server_name` ]**<br>
This is what you want your FiveM server to be called, this value is used in various places as an auto-fill like bot status and commands as well as translations when people are connecting to your server.

### DiscordInviteLink
**[ Replicated | Convar: `discord_invite` ]**<br>
A link to your discord server that will be given to users on connection problems to the FiveM server, used for discord presence links and some command replies.

### FiveMServerIP
**[ Replicated | Convar: `discord_server_ip` ]**<br>


### DebugLogs
Turns on debug logging from discord.js for checking for errors
**[ Replicated | Convar: `discord_debug` ]**<br>

# Discord Bot Settings

### EnableDiscordBot
**[ Convar: `discord_enable_bot`]**<br>
This option deterrmins whether the bot actually connects to discord at all. if this is set to false the bot portions will not work at all. but log webhooks, presence updates and connection queue will continue though without priority


## Discord Bot

### DiscordBotToken
**[ Convar: `discord_token`]**<br>
This is where you put your discord bot token generated from [Discord Applications](https://discord.com/developers/applications). It will look somewhat like `ZDkx5NjQxjExNMzDM1D2U0O4.v9g7Wa.DK700Xi2LeVOC2NUuABgPt0ZyWR`.
<br>**DO NOT** Share this token with anyone, it's a key to your bot and discord server which could be dangerous in the wrong hands.

### DiscordGuildId
**[ Convar: `discord_guild_id` ]**<br>
This is for your Discord Server id (Also called a Guild Id). You can get that by enabling developer mode in your discord client then right clicking your discord server icon and "copy id". This is an important value, the bot will only listen to events done within this server.


## Staff Chat

### EnableStaffChatForwarding
**[ Convar: `discord_enable_staff_chat`]**<br>
This set to "true" will enable the feature of forwarding chat from in game to discord and back in the channel configured in the next setting.

### DiscordStaffChannelId
**[ Convar: `discord_staff_channel_id` ]**<br>
This is the channel staff chat will be sent and be taken from to send to staff in game. this can only be 1 channel and some things might not look the same in game as they do in discord like emoji or @mentions but that's normal.

### AdditionalStaffChatRoleIds
if you have extra roles you'd like to access staff chat other than the default mod, admin and god roles configured below you can add other roles here and they'll have access to see and toggle staff chat. Just make sure they can also see the DiscordStaffChannelId channel ;)


## Whitelisting / Allowlisting

### EnableWhitelistChecking
**[ Convar: `discord_enable_whitelist` ]**<br>
This setting determines whether the bot should check if someone is in the discord and if they have a preset role defined under [DiscordWhitelistRoleIds](#DiscordWhitelistRoleIds). Setting this to `"true"` wont let anyone into the server unless they:

1. Have Discord open.
2. Are in the Discord defined in [DiscordGuildId](#DiscordGuildId).
3. Have the whitelisted role in the discord

### DiscordWhitelistRoleIds
**[ Convar: `discord_whitelist_roles` ]**<br>
This is a comma separated list (`"role1id, role2id, role3id"`) of role IDs that will be able to connect to the server if [EnableWhitelistChecking](#EnableWhitelistChecking) is set to "true". To get a role Id you enable developer mode in your discord client then either right click a role name from someone's popup profile or right click the role name in the role settings and "copy id".


## Slash Commands / Discord Permissions

### EnableDiscordSlashCommands
**[ Convar: `discord_enable_commands` ]**<br>
If this is set to `"true"` the bot will register all it's slash commands to the discord server and allow users with the correct roles to access them. if this is set to `"false"` all the commands will ignored and never registered.

### DiscordModRoleId
**[ Convar: `discord_mod_role` ]**<br>
This is a single discord role ID that will be permitted to use commands set to role `"mod"`. this can only be a single role unlike [DiscordWhitelistRoleIds](#DiscordWhitelistRoleIds). You can add other roles but you'll have to follow the [guide](https://zfbx.github.io/zdiscord/commands#add-permission-levels). You can get the role ID in the same way described under [DiscordWhitelistRoleIds](#DiscordWhitelistRoleIds)

### DiscordAdminRoleId
**[ Convar: `discord_admin_role` ]**<br>
This is a single discord role ID that will be permitted to use commands set to role `"admin"` OR `"mod"`. Otherwise refer to [DiscordModRoleId](#DiscordModRoleId)'s notes.

### DiscordGodRoleId
**[ Convar: `discord_god_role` ]**<br>
This is a single discord role ID that will be permitted to use **all** commands. Including `"admin"` and `"mod"`. Otherwise refer to [DiscordModRoleId](#DiscordModRoleId)'s notes.


## Discord Bot Status

### EnableBotStatusMessages
**[ Convar: `discord_enable_status` ]**<br>
If this is set to `"true"` the bot will pick a random message from the [BotStatusMessages](#BotStatusMessages) config and display it as it's status. (Example: "Playing FiveM")

### BotStatusMessages
The bot will pick a random message from this array every 30 seconds to set as it's status if [EnableBotStatusMessages](#EnableBotStatusMessages) is set to `"true"`. you must have at least 1 of these if enabled. You can use `{playercount}`, `{servername}` or `{invite}` inside the status messages and when they're shown they will replace those values with either the current number of players online, the server name set under [FiveMServerName](#FiveMServerName) or invite set under [DiscordInviteLink](#DiscordInviteLink).


## Ace Permissions

### EnableAutoAcePermissions
**[ Convar: `discord_enable_ace_perms` ]**<br>
If this is set to `"true"` it'll enable a system of granting users Ace Permissions based on their role configured under [AutoAcePermissions](#AutoAcePermissions).

**NOTE: This will only work if `add_ace resource.zdiscord command allow` is set in the server.cfg**

### AutoAcePermissions
If [EnableAutoAcePermissions](#EnableAutoAcePermissions) is set to `"true"` the configured ace permissions will be automatically given to users which have at least one of the matching roles. For example if you put `"group.police": "DiscordPoliceRoleID",` as one of the rows, if a user joined the server and had the `DiscordPoliceRoleID` in discord the bot will automatically set the `group.police` ace permission to them. On disconnect all ace permissions granted are removed.

By default the discord mod, admin and god roles are given `group.mod`, `group.admin` or `group.god` respectively.

Set a single role id with `"example": "000000000000000000",`. You can also have multiple discord roles checked for an Ace permission by using an array instead of a single roleId string like: `"example2": [ "000000000000000000", "000000000000000000"],`

**NOTE: This will only work if `add_ace resource.zdiscord command allow` is set in the server.cfg**


## Other

### saveScreenshotsToServer
**[ Convar: `discord_save_screenshots` ]**<br>
If you want discord `/screenshot`s to be saved locally to the server set this to `"true"` however if you just want the screenshots to be sent to discord when the command is used leave this as `"false"`. Please note if you change this to true, it's up to you to monitor your screenshots folder as images can be quite large and use a lot of space if they don't get cleaned up regularly.



# Webhook Logging Settings

### EnableLoggingWebhooks
**[ Convar: `discord_enable_logging_webhooks` ]**<br>
This setting enables the abilty to use the log export. if this is set to false the export will return false and not send the event.

### LoggingWebhookName
**[ Convar: `discord_logging_name` ]**<br>
This is the name to display in discord when the logs come through. Webhooks work slightly different from the normal bot and can have their own names or pictures set without issue, these "bots" however can't read messages or show up in the user list, it's just a one way message

### LoggingAlertPingId
**[ Convar: `discord_logging_ping_id` ]**<br>
This discord id will be pingged in the event a log is sent with `true` for pingRole. This can be used to give attention to an important log that a dedicated staff team might want to know about or deal with for example a monetary transaction over a certain amount might be suspicious and warrent a ping to this user or role. There can only be one id for this.<br>
**IMPORTANT NOTE:** If you want to ping a user just put the ID, if you want to **ping a role** put a `&` in front of the role id**

### LoggingWebhooks
This is the array of webhooks you setup to be used with the log system. it uses `"key": "value",` where `key` is the name (case sensitive) of the webhook for you to be able to call from the logging export as the first value




docs/changelog.md
## Change log

**7.2.0 - Addon improvements [2022-01-16]**

- patched some errors from previous version
- added example events for hooking from discord
- updated api usage to be stable with newest branch of discord.js
- added config for logging debug messages from discord.js

**7.1.0 - Addon improvements [2022-01-14]**

- added Italian from anonymous
- bug fixes
- add Czech from Anostraca
- add vehicle commands
- updated oxmysql compatibility
- add Norwegian from VikingTheDev
- optional addon for live player count added
- Updated dependencies removing outdated and improving stablility

**7.0.3 - Fixes + New config for addl staffchat roles [2021-12-10]**

- fixed a few logging typos
- added new config `AdditionalStaffChatRoleIds` (read config docs for more info)

**7.0.2 - New languages and hot fixes [2021-12-9]**

- added Bulgarian from @WrecksBG
- added Spanish from @Xect0r
- Fix /message error

**7.0.1 - New language and QoL [2021-12-8]**

- added French from @tiweb442
- add sendMessage chat util

**7.0.0 - Staff chat, new config AND WEBHOOK LOGGING [2021-12-8]**

- added bi-directional staff chat (messages from discord staff channel <-> fiveM staff chat messages)
- add discord webhook logging export (think qb-logs)
- Refactored and added to config to be more clear
- dutch added. Thanks Chatty <3
- Slovenian added. Thanks Synthetics <3
- remove command versioning
- converted bot into class
- added addon support
- relicensed under CC-BY-SA-NC-4.0

**6.0.0 - Refactor! [2021-10-25]**

- command arguments have been refactored to pass as an object instead of an array (This is not backwards compatible which is why it's a version bump)

**5.0.2 - New commands and docs! [2021-10-23]**

- added screenshot command
- added weather, blackout, time commands
- added docs at [zfbx.github.io/zdiscord](https://zfbx.github.io/zdiscord)
- updated/fixed chat events and teleport command

**5.0.1 - Exports and ace perms [2021-10-20]**

- Automatic Ace permission granting by discord role
- new and changed exports for checking if role exist and getting roles

**5.0.0 - Hello discord.js! & Much much more. [2021-10-18]**

*It should be noted, this is essentially a complete rewrite of the resource and should probably be treated as such.*

- NEW LICENSE - GPL version 3.0
- Replaced eris with discord.js
- replaced commands with slash commands
- added a bunch of new commands
- added a utils file to cleanup the main script
- added client.js for running events client side
- added linting and editorconfig

**4.0.0 - permissions! bye yarn!.. and esx. [2021-9-15]**

- Dependencies are now embedded libs making the project slimmer on install and not requiring yarn.
- ESX support was dropped, it wasn't getting used and I don't have an esx server for testing.
- translations were slimmed down to fivem messages only.
- `add/removepermissions` merged into `perms` command
-  Convars changed to be more readable using standard FiveM format with `_` spacing

**3.2.0 - Convars everywhere! [2021-7-24]**

- Convar hooks added for nearly every config option [Read More](https://github.com/zfbx/zdiscord/wiki/Convars)
- Slimmed up locales by removing console log messages
- Added {prefix} to the global variables
- New Add/Remove permissions command for QBCore


**3.1.0 - QBCore (potential ESX) support added [2021-7-10]**

- Commands starting with `qb-` will load automatically if QBCore is detected.
- Placeholders for `esx-` commands have been added.


**3.0.0 - Modular commands! [2021-7-10]**

- Commands are now loaded dynamically from the `/commands` folder
- Help command now has sub commands `!help commandName`
- New DM command
- various Error checks and fixes


**2.0.0 - Github Release [2021-7-9]**

- polished standalone
- Added translation support


**1.0.0 - First unsupported build [2021-6-10]**

- A lot.


config.js
/*
    zdiscord - by Tony/zfbx - https://github.com/zfbx/zdiscord - License: CC BY-NC-SA 4.0
    Docs for this file available at https://zfbx.github.io/zdiscord/config or in docs/config.md
*/


/** ******************************
 * GENERAL CONFIGURATION SETTINGS
 ********************************/

const LanguageLocaleCode = "en";

// PUBLIC VALUES
const FiveMServerName = "My FiveM Server";
const DiscordInviteLink = "https://discord.gg/fivem";
const FiveMServerIP = "127.0.0.1";

// This spams the console, only enable for testing if needed
const DebugLogs = false;


/** ********************
 * DISCORD BOT SETTINGS
 ***********************/

const EnableDiscordBot = true;

// DISCORD BOT
const DiscordBotToken = "CHANGE";
const DiscordGuildId = "000000000000000000";

// STAFF CHAT
const EnableStaffChatForwarding = false;
const DiscordStaffChannelId = "000000000000000000";
const AdditionalStaffChatRoleIds = [
    // "000000000000000",
];

// WHITELISTING / ALLOWLISTING
const EnableWhitelistChecking = true;
const DiscordWhitelistRoleIds = "000000000000000000, 000000000000000000";

// SLASH COMMANDS / DISCORD PERMISSIONS
const EnableDiscordSlashCommands = true;
const DiscordModRoleId = "000000000000000000";
const DiscordAdminRoleId = "000000000000000000";
const DiscordGodRoleId = "000000000000000000";

// DISCORD BOT STATUS
const EnableBotStatusMessages = true;
const BotStatusMessages = [
    "{servername}",
    "{playercount} online",
];

// ACE PERMISSIONS
const EnableAutoAcePermissions = false;
const AutoAcePermissions = {
    // "example": "000000000000000000",
    // "example2": [ "000000000000000000", "000000000000000000"],
};

// Other
const SaveScreenshotsToServer = false;


/** ************************
 * WEBHOOK LOGGING SETTINGS
**************************/

const EnableLoggingWebhooks = false;
const LoggingWebhookName = "zLogs";
// put "&" in front of the id if you're to ping a role
const LoggingAlertPingId = "&000000000000000000";
// example: "bank": "https://discord.com/webhook/...",
const LoggingWebhooks = {
    "example": "https://discord.com/api/webhooks/000000000/sEcRRet-ToK3n_5tUfF_tH8t_YUo-S40u1d-n07-sHar3",
};


/** !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
 * !! DO NOT EDIT BELOW THIS LINE UNLESS YOU KNOW WHAT YOU'RE DOING !!
 * !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!*/

module.exports = {
    EnableDiscordBot: getConBool("discord_enable_bot", EnableDiscordBot),
    EnableStaffChatForwarding: getConBool("discord_enable_staff_chat", EnableStaffChatForwarding),
    EnableLoggingWebhooks: getConBool("discord_enable_logging_webhooks", EnableLoggingWebhooks),
    DebugLogs: getConBool("discord_debug", DebugLogs),
    DiscordBotToken: GetConvar("discord_token", DiscordBotToken),
    DiscordGuildId: GetConvar("discord_guild_id", DiscordGuildId),
    LanguageLocaleCode: GetConvar("discord_lang", LanguageLocaleCode),
    FiveMServerName: GetConvar("discord_server_name", FiveMServerName),
    DiscordInviteLink: GetConvar("discord_invite", DiscordInviteLink),
    FiveMServerIP: GetConvar("discord_server_ip", FiveMServerIP),
    EnableWhitelistChecking: getConBool("discord_enable_whitelist", EnableWhitelistChecking),
    DiscordWhitelistRoleIds: getConList("discord_whitelist_roles", DiscordWhitelistRoleIds),
    EnableDiscordSlashCommands: getConBool("discord_enable_commands", EnableDiscordSlashCommands),
    DiscordModRoleId: GetConvar("discord_mod_role", DiscordModRoleId),
    DiscordAdminRoleId: GetConvar("discord_admin_role", DiscordAdminRoleId),
    DiscordGodRoleId: GetConvar("discord_god_role", DiscordGodRoleId),
    EnableBotStatusMessages: getConBool("discord_enable_status", EnableBotStatusMessages),
    BotStatusMessages: BotStatusMessages,
    EnableAutoAcePermissions: getConBool("discord_enable_ace_perms", EnableAutoAcePermissions),
    AutoAcePermissions: AutoAcePermissions,
    SaveScreenshotsToServer: getConBool("discord_save_screenshots", SaveScreenshotsToServer),
    DiscordStaffChannelId: GetConvar("discord_staff_channel_id", DiscordStaffChannelId),
    LoggingWebhooks: LoggingWebhooks,
    LoggingAlertPingId: GetConvar("discord_logging_ping_id", LoggingAlertPingId),
    LoggingWebhookName: GetConvar("discord_logging_name", LoggingWebhookName),
    StaffChatRoleIds: [
        GetConvar("discord_mod_role", DiscordModRoleId),
        GetConvar("discord_admin_role", DiscordAdminRoleId),
        GetConvar("discord_god_role", DiscordGodRoleId),
        ...AdditionalStaffChatRoleIds,
    ],
};

/** Returns convar or default value fixed to a true/false boolean
 * @param {boolean|string|number} con - Convar name
 * @param {boolean|string|number} def - Default fallback value
 * @returns {boolean} - parsed bool */
function getConBool(con, def) {
    if (typeof def == "boolean") def = def.toString();
    const ret = GetConvar(con, def);
    if (typeof ret == "boolean") return ret;
    if (typeof ret == "string") return ["true", "on", "yes", "y", "1"].includes(ret.toLocaleLowerCase().trim());
    if (typeof ret == "number") return ret > 0;
    return false;
}

/** returns array of items or default array provided
 * @param {string} con - string of comma separated values
 * @param {string|Array} def - string of comma separated values
 * @returns {object} - array of discord ids */
function getConList(con, def) {
    const ret = GetConvar(con, def);
    if (typeof ret == "string") return ret.replace(/[^0-9,]/g, "").replace(/(,$)/g, "").split(",");
    if (Array.isArray(ret)) return ret;
    if (!ret) return [];
}
